### PR TITLE
New command port-forward: port-forward TCP and UDP ports from the device

### PR DESCRIPTION
**Pull request by tranchitella**
Draft PR, looking for preliminary comments/feedbacks.

### CODE REVIEW

File: client/deviceconnect/client.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+	return &Client{
+		url:        url,
+		skipVerify: skipVerify,
+	}
+}
+
+// Connect to the websocket
+func (c *Client) Connect(deviceID string, token []byte) error {
+	fmt.Printf("Connecting to the device %s...\n", deviceID)
+	deviceConnectPath := "/api/management/v1/deviceconnect/devices/" + deviceID + "/connect"
```
Comment (merlin-northern) [inline client/deviceconnect/client.go:60]: can we have a const/var for that?

File: cmd/portforward.go [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+const (
+	propertyConnectionID = "connection_id"
+	propertyRemoteHost   = "remote_host"
+	propertyRemotePort   = "remote_port"
+	propertyProtocol     = "protocol"
+)
+
+var portForwardCmd = &cobra.Command{
+	Use:   "port-forward DEVICE_ID LOCAL_PORT:REMOTE_PORT ...",
+	Short: "Forward one or more local ports to a device",
```
Comment (merlin-northern) [inline cmd/portforward.go:58]: suggested rephrase: the description could be better, where is the local port open and where the remote one, what is forwarded to what.

File: cmd/portforward.go [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+
+const (
+	propertyConnectionID = "connection_id"
+	propertyRemoteHost   = "remote_host"
+	propertyRemotePort   = "remote_port"
+	propertyProtocol     = "protocol"
+)
+
+var portForwardCmd = &cobra.Command{
+	Use:   "port-forward DEVICE_ID LOCAL_PORT:REMOTE_PORT ...",
```
Comment (merlin-northern) [inline cmd/portforward.go:57]: my dreams: that we follow the standard man-like options description schema:
```
SYNOPSIS
     tar [bundled-flags <args>] [<file> | <pattern> ...]
```

File: cmd/portforward.go [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+	Args:  cobra.MinimumNArgs(2),
+	Run: func(c *cobra.Command, args []string) {
+		cmd, err := NewPortForwardCmd(c, args)
+		CheckErr(err)
+		CheckErr(cmd.Run())
+	},
+}
+
+func init() {
+	portForwardCmd.Flags().StringP(argBindHost, "", "127.0.0.1", "binding host")
```
Comment (merlin-northern) [inline cmd/portforward.go:68]: can we have 127.0.0.1 as a const/variable?

File: cmd/portforward.go [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+	running      bool
+	stop         chan struct{}
+	err          error
+}
+
+func getPortMappings(args []string) ([]portMapping, error) {
+	var err error
+	portMappings := []portMapping{}
+	for _, arg := range args {
+		remoteHost := "127.0.0.1"
```
Comment (merlin-northern) [inline cmd/portforward.go:100]: ssh -R  support -- so close!  :)

File: cmd/portforward.go [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+}
+
+type TCPPortForwarder struct {
+	listen     net.Listener
+	remoteHost string
+	remotePort uint
+}
+
+func NewTCPPortForwarder(bindingHost string, localPort uint, remoteHost string, remotePort uint) (*TCPPortForwarder, error) {
+	fmt.Printf("Forwarding from %s:%d -> %s:%d\n", bindingHost, localPort, remoteHost, remotePort)
```
Comment (merlin-northern) [inline cmd/portforward.go:198]: do we really have no logger in the cli? (sorry have not checked)

File: cmd/portforward.go [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+
+func (c *PortForwardCmd) processIncomingMessages(msgChan chan *ws.ProtoMsg, client *deviceconnect.Client) {
+	for c.running {
+		m, err := client.ReadMessage()
+		if err != nil {
+			if c.running {
+				fmt.Fprintf(os.Stderr, "error: %v\n", err)
+			} else {
+				c.Stop()
+			}
```
Comment (merlin-northern) [inline cmd/portforward.go:529]: I do not get it, if it is running, then we output error, and in other case we stop?

**Review by merlin-northern — COMMENTED** [review]
looks good, some comments, one question. I trust your idea of implementing it.
so, what about full tests for it? :>

File: client/deviceconnect/client.go (hunk 1 above)
Comment (tranchitella) [reply client/deviceconnect/client.go:60]: done

File: cmd/portforward.go (hunk 2 above)
Comment (tranchitella) [reply cmd/portforward.go:58]: Improved the help, including examples, thanks for your suggestion!

File: cmd/portforward.go (hunk 4 above)
Comment (tranchitella) [reply cmd/portforward.go:68]: done

File: cmd/portforward.go (hunk 5 above)
Comment (tranchitella) [reply cmd/portforward.go:100]: not really, this is more like `-L 8080:192.168.1.1:8080`, and it works :)

File: cmd/portforward.go (hunk 6 above)
Comment (tranchitella) [reply cmd/portforward.go:198]: no logger

File: cmd/portforward.go (hunk 7 above)
Comment (tranchitella) [reply cmd/portforward.go:529]: yes, we receive a spurious error when we stop (e.g. io.EOF), and we simply ignore it in this case

File: cmd/portforward.go (hunk 3 above)
Comment (tranchitella) [reply cmd/portforward.go:57]: Done

**Comment by tranchitella** [comment]
> looks good, some comments, one question. I trust your idea of implementing it.
> so, what about full tests for it? :>

mender-cli doesn't have unit tests, there are acceptance tests but without a running device.
Maybe we should combine mender-cli and integration test (which we'll need anyway) and do the full testing in integration.
What do you think?

**Comment by merlin-northern** [comment]
> > looks good, some comments, one question. I trust your idea of implementing it.
> > so, what about full tests for it? :>
> 
> mender-cli doesn't have unit tests, there are acceptance tests but without a running device.
> Maybe we should combine mender-cli and integration test (which we'll need anyway) and do the full testing in integration.
> What do you think?

makes sense to me, like include mender-cli in the full integration tests. but it can be big. I think we can release without it.

File: cmd/portforward.go [hunk 8]
```diff
@@ -0,0 +1,560 @@
...
+)
+
+const (
+	argBindHost    = "bind"
+	readBuffLength = 64 * 1024
+	localhost      = "127.0.0.1"
+)
+
+const (
+	ProtoTypePortForward       = 3
```
Comment (kacf) [inline cmd/portforward.go:45]: Similar to mender-connect, perhaps these Proto Types should be grouped.

File: cmd/portforward.go [hunk 9]
```diff
@@ -0,0 +1,560 @@
...
+			}
+			arg = parts[1]
+		}
+		var localPort, remotePort int
+		if strings.Contains(arg, ":") {
+			parts := strings.SplitN(arg, ":", 3)
+			if len(parts) == 3 {
+				remoteHost = parts[0]
+				parts = parts[1:]
+			}
```
Comment (kacf) [inline cmd/portforward.go:131]: +1 for inclusion of host in the port specification!

However, this is the opposite order of the one used in ssh, e.g. `remoteserver.com:8080:80` vs `8080:remoteserver.com:80`. Is it better to swap them so that we are command line compatible?

SSH also has `10.0.0.10:8080:80`, but it means "bind to 10.0.0.10", so quite different meaning.

File: cmd/portforward.go (hunk 8 above)
Comment (tranchitella) [reply cmd/portforward.go]: they moved to go-lib-micro

File: cmd/portforward.go (hunk 9 above)
Comment (tranchitella) [reply cmd/portforward.go:121]: fixed, it is now `LOCAL_PORT:REMOTE_HOST:REMOTE_PORT`, with a simplified version `LOCAL_PORT:REMOTE_PORT`

File: cmd/portforward.go [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+	localhost      = "127.0.0.1"
+)
+
+var portForwardCmd = &cobra.Command{
+	Use:   "port-forward [DEVICE_ID] [LOCAL_PORT:REMOTE_PORT] ...",
+	Short: "Forward a local port to a remote port on the device",
+	Long: "This command supports both TCP and UDP port-forwarding.\n\n" +
+		"The port specification can be prefixed with \"tcp/\" or \"udp/\".\n" +
+		"If no prefix is specified, TCP is the default.\n\n" +
+		"REMOTE_PORT can also be specified in the form REMOTE_HOST:REMOT_PORT, making\n" +
```
Comment (kacf) [inline cmd/portforward.go:49]: `s/REMOT_PORT/REMOTE_PORT/`

**Review by kacf — APPROVED** [review]
Looks good to me (one typo below).

Maybe @alfrunes wants to take a small look at the protocol part, just to be sure.

File: client/deviceconnect/client.go [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// ReadMessage reads a Proto message from the websocket
+func (c *Client) ReadMessage() (*ws.ProtoMsg, error) {
+	_, data, err := c.conn.ReadMessage()
```
Comment (alfrunes) [inline client/deviceconnect/client.go:130]: `c.conn.Read*` should be protected by a mutex: https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency

File: client/deviceconnect/client.go [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	m := &ws.ProtoMsg{}
+	err = msgpack.Unmarshal(data, m)
+	if err != nil {
+		return nil, err
+	}
```
Comment (alfrunes) [inline client/deviceconnect/client.go:139]: End-to-end (ProtoMsg) [ping/pong](https://github.com/mendersoftware/go-lib-micro/blob/master/ws/protomsg.go#L45) is not handled anywhere.

File: client/deviceconnect/client.go [hunk 13]
```diff
@@ -0,0 +1,161 @@
...
+// WriteMessage writes a Proto message to the websocket
+func (c *Client) WriteMessage(m *ws.ProtoMsg) error {
+	data, err := msgpack.Marshal(m)
+	if err != nil {
+		return errors.Wrap(err, "Unable to marshal the message from the websocket")
+	}
+	if err := c.conn.SetWriteDeadline(time.Now().Add(writeWait)); err != nil {
+		return errors.Wrap(err, "Unable to set the write deadline")
+	}
+	if err := c.conn.WriteMessage(websocket.BinaryMessage, data); err != nil {
```
Comment (alfrunes) [inline client/deviceconnect/client.go:152]: `c.conn.Write*` should be protected by a mutex: https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency

File: client/deviceconnect/client.go (hunk 11 above)
Comment (tranchitella) [reply client/deviceconnect/client.go:130]: @alfrunes  I can add the mutex, but I believe there is a single go-routine reading and writing messages using the gorilla websocket connection. Am I wrong? By the way, I simply refactored the previous implementation from the Remote Terminal into a dedicated client to reuse it in the Port Forwarding, I'm not changing the RT implementation.

File: client/deviceconnect/client.go (hunk 13 above)
Comment (tranchitella) [reply client/deviceconnect/client.go:152]: See comment above about "Read"

File: cmd/artifact_upload.go [hunk 14]
```diff
@@ -29,7 +29,7 @@ const (
 )
 
 var artifactUploadCmd = &cobra.Command{
-	Use:   "upload [flags] ARTIFACT",
+	Use:   "upload [flags] [ARTIFACT]",
```
Comment (lluiscampos) [inline cmd/artifact_upload.go:32]: Why this change? We are not making `ARTIFACT` optional.

File: cmd/portforward.go [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+)
+
+const (
+	argBindHost    = "bind"
+	readBuffLength = 4096
+	localhost      = "127.0.0.1"
+)
+
+var portForwardCmd = &cobra.Command{
+	Use:   "port-forward [DEVICE_ID] [LOCAL_PORT:REMOTE_PORT] ...",
```
Comment (lluiscampos) [inline cmd/portforward.go:44]: `DEVICE_ID` is not optional, to be consistent with the rest of the tool it should be:
```suggestion
	Use:   "port-forward DEVICE_ID [LOCAL_PORT:REMOTE_PORT] ...",
```

or even better something like:
```suggestion
	Use:   "port-forward DEVICE_ID [tcp|udp/]LOCAL_PORT[:REMOTE_PORT] [[tcp|udp/]LOCAL_PORT[:REMOTE_PORT]...]",
```

File: cmd/portforward.go [hunk 16]
```diff
@@ -0,0 +1,384 @@
...
+	Short: "Forward a local port to a remote port on the device",
```
Comment (lluiscampos) [inline cmd/portforward.go:45]:
```suggestion
	Short: "Forward one or more local ports to remote port(s) on the device",
```

File: cmd/artifact_upload.go (hunk 14 above)
Comment (tranchitella) [reply cmd/artifact_upload.go:32]: I'm sorry, it's a mistake

File: client/deviceconnect/client.go (hunk 12 above)
Comment (tranchitella) [reply client/deviceconnect/client.go:139]: I handled the "pong" when receiving ping. I don't think we have the need of sending the "pings" in the CLI tool, what do you think?

Comment (alfrunes) [reply client/deviceconnect/client.go:139]: I agree, it's not necessary.

File: client/deviceconnect/client.go (hunk 11 above)
Comment (alfrunes) [reply client/deviceconnect/client.go:130]: websocket ping packet is written in a separate go-routine (actually the routine calling ReadMessage). I'm not sure if the packet is so small that can be considered to be a single "atomic" write, but I'd rather be safe than sorry.

Comment (tranchitella) [reply client/deviceconnect/client.go:137]: You are right, let's do it: https://github.com/mendersoftware/mender-cli/pull/106/commits/243c0b7c73650314f9a2e6e371874fc8130bcb6a
